keras_nlp/models/gpt2.py

```diff
@@ -0,0 +1,287 @@
+            shape=(None,), dtype="int32", name="padding_mask"
+        )
+
+        # Embed tokens, positions.
+        token_embedding = keras.layers.Embedding(
+            input_dim=vocabulary_size,
+            output_dim=hidden_dim,
+            embeddings_initializer=_gpt_2_kernel_initializer(stddev=0.01),
+            name="token_embedding",
+        )(token_ids)
+
+        position_embedding = PositionEmbedding(
+            initializer=_gpt_2_kernel_initializer(stddev=0.02),
+            sequence_length=max_sequence_length,
+            name="position_embedding",
+        )(token_embedding)
+
+        # Sum and apply dropout to embeddings.
+        x = keras.layers.Add()((token_embedding, position_embedding))
+        x = keras.layers.Dropout(
+            dropout,
+            name="embeddings_dropout",
+        )(x)
+
+        # Apply successive transformer decoder blocks.
+        for i in range(num_layers):
+            x = TransformerDecoder(
+                intermediate_dim=intermediate_dim,
+                num_heads=num_heads,
+                dropout=dropout,
+                activation=lambda x: keras.activations.gelu(
+                    x, approximate=True
+                ),
+                layer_norm_epsilon=1e-05,
+                kernel_initializer=_gpt_2_kernel_initializer(stddev=0.02),
+                normalize_first=True,
+                name=f"transformer_layer_{i}",
+            )(x, decoder_padding_mask=padding_mask)
+
+        sequence_output = keras.layers.LayerNormalization(
+            name="layer_norm",
+            axis=-1,
+            epsilon=1e-05,
+            dtype=tf.float32,
+        )(x)
+
+        # Instantiate using Functional API Model constructor
+        super().__init__(
+            inputs={
+                "token_ids": token_ids,
+                "padding_mask": padding_mask,
+            },
+            outputs=sequence_output,
+            name=name,
+            trainable=trainable,
+        )
+        # All references to `self` below this line
+        self.vocabulary_size = vocabulary_size
+        self.num_layers = num_layers
+        self.num_heads = num_heads
+        self.hidden_dim = hidden_dim
+        self.intermediate_dim = intermediate_dim
+        self.dropout = dropout
+        self.max_sequence_length = max_sequence_length
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "vocabulary_size": self.vocabulary_size,
+                "num_layers": self.num_layers,
+                "num_heads": self.num_heads,
+                "hidden_dim": self.hidden_dim,
+                "intermediate_dim": self.intermediate_dim,
+                "dropout": self.dropout,
+                "max_sequence_length": self.max_sequence_length,
+            }
+        )
+        return config
+
+
+MODEL_DOCSTRING = """GPT-2 implementation using "{type}"
+    architecture (with {num_params} parameters).
```

Comment:
Changed it to `GPT-2 core network with "{type}" architecture ({num_params} parameters).` after addressing Matt's comment. Two "with"s might look weird in the same sentence

Reply:
On second thoughts, changed it to 

`GPT-2 "{type}" architecture with {num_params} parameters.`

Reply:
I really like the parameter count as a usability improvement, but I'm not sure we can do this totally accurately right?

vocabulary_size=10 vs vocabulary_size=1_000_000 would kinda invalidate your parameter count. I think we need to do this at the checkpoint level. Or say "roughly XXX parameters".

Reply:
Let's just remove this for now, and plan to document parameter count with actual checkpoints.

Reply:
@mattdangerw, the reason why I put #params in the description is because that is how the official repo distinguishes between the different GPT-2 sizes: https://github.com/openai/gpt-2/blob/master/model_card.md. Even the paper does not mention "Base", "Medium", "Large", "ExtraLarge". However, since changing the vocabulary_size will change #param, I'll remove #param altogether, makes sense to have it at the ckpt level.

Reply:
Yeah totally. I am torn because it's very useful.

But GPT which is only releasing a single checkpoint so didn't have to deal with the general case here. And the parameter count difference between checkpoints can get pretty big.

Funnily enough, the original bert repo is actually falling into this trap. It is claiming the multi lingual version is 110M params when its almost 180. There's even a poor forgotten PR for it https://github.com/google-research/bert/pull/1150